ipalib/constants.py

```diff
@@ -35,6 +35,28 @@
     except Exception:
         FQDN = None
 
+# TLS related constants
+# * SSL2 and SSL3 are broken.
+# * TLS1.0 and TLS1.1 are no longer state of the art.
+# * TLS1.3 support is not yet stable, e.g. issues with PHA.
+# Therefore only TLS 1.2 is enabled by default.
+
+TLS_VERSIONS = [
+    "ssl2",
+    "ssl3",
+    "tls1.0",
+    "tls1.1",
+    "tls1.2",
+    "tls1.3",
+]
+TLS_VERSION_MINIMAL = "tls1.0"
+TLS_VERSION_DEFAULT_MIN = "tls1.2"
+TLS_VERSION_DEFAULT_MAX = "tls1.2"
+
+# high ciphers without RC4, MD5, TripleDES, pre-shared key
+# and secure remote password
+TLS_HIGH_CIPHERS = "HIGH:!aNULL:!eNULL:!MD5:!RC4:!3DES:!PSK:!SRP"
```

Comment:
Is this cipher string still used? It enables a bunch of ciphers with weak algorithms like SHA1, ARIA, and DSS. The [SSL Configuration Generator](https://ssl-config.mozilla.org/#server=apache&server-version=2.4.39&config=intermediate&openssl-version=1.0.2) recommends ``ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384`` as cipher string.

Reply:
In mod_nss this will only enable SHA-1 and not ARIA or DSS. The NSS cipher list is:

aes_128_sha_256, aes_256_sha_256, ecdhe_ecdsa_aes_128_gcm_sha_256, ecdhe_ecdsa_aes_128_sha, ecdhe_ecdsa_aes_256_gcm_sha_384, ecdhe_ecdsa_aes_256_sha, ecdhe_rsa_aes_128_gcm_sha_256, ecdhe_rsa_aes_128_sha, ecdhe_rsa_aes_256_gcm_sha_384, ecdhe_rsa_aes_256_sha, rsa_aes_128_gcm_sha_256, rsa_aes_128_sha, rsa_aes_256_gcm_sha_384, rsa_aes_256_sha

Reply:
The string is used to configure an OpenSSL ``ssl.SSLContext`` object, not NSS:

```
>>> import ssl
>>> ctx = ssl.SSLContext()
>>> ctx.set_ciphers("HIGH:!aNULL:!eNULL:!MD5:!RC4:!3DES:!PSK:!SRP")
>>> list(c['name'] for c in ctx.get_ciphers())
['TLS_AES_256_GCM_SHA384', 'TLS_CHACHA20_POLY1305_SHA256', 'TLS_AES_128_GCM_SHA256', 'TLS_AES_128_CCM_SHA256', 'ECDHE-ECDSA-AES256-GCM-SHA384', 'ECDHE-RSA-AES256-GCM-SHA384', 'DHE-DSS-AES256-GCM-SHA384', 'DHE-RSA-AES256-GCM-SHA384', 'ECDHE-ECDSA-CHACHA20-POLY1305', 'ECDHE-RSA-CHACHA20-POLY1305', 'DHE-RSA-CHACHA20-POLY1305', 'ECDHE-ECDSA-AES256-CCM8', 'ECDHE-ECDSA-AES256-CCM', 'DHE-RSA-AES256-CCM8', 'DHE-RSA-AES256-CCM', 'ECDHE-ECDSA-ARIA256-GCM-SHA384', 'ECDHE-ARIA256-GCM-SHA384', 'DHE-DSS-ARIA256-GCM-SHA384', 'DHE-RSA-ARIA256-GCM-SHA384', 'ECDHE-ECDSA-AES128-GCM-SHA256', 'ECDHE-RSA-AES128-GCM-SHA256', 'DHE-DSS-AES128-GCM-SHA256', 'DHE-RSA-AES128-GCM-SHA256', 'ECDHE-ECDSA-AES128-CCM8', 'ECDHE-ECDSA-AES128-CCM', 'DHE-RSA-AES128-CCM8', 'DHE-RSA-AES128-CCM', 'ECDHE-ECDSA-ARIA128-GCM-SHA256', 'ECDHE-ARIA128-GCM-SHA256', 'DHE-DSS-ARIA128-GCM-SHA256', 'DHE-RSA-ARIA128-GCM-SHA256', 'ECDHE-ECDSA-AES256-SHA384', 'ECDHE-RSA-AES256-SHA384', 'DHE-RSA-AES256-SHA256', 'DHE-DSS-AES256-SHA256', 'ECDHE-ECDSA-CAMELLIA256-SHA384', 'ECDHE-RSA-CAMELLIA256-SHA384', 'DHE-RSA-CAMELLIA256-SHA256', 'DHE-DSS-CAMELLIA256-SHA256', 'ECDHE-ECDSA-AES128-SHA256', 'ECDHE-RSA-AES128-SHA256', 'DHE-RSA-AES128-SHA256', 'DHE-DSS-AES128-SHA256', 'ECDHE-ECDSA-CAMELLIA128-SHA256', 'ECDHE-RSA-CAMELLIA128-SHA256', 'DHE-RSA-CAMELLIA128-SHA256', 'DHE-DSS-CAMELLIA128-SHA256', 'ECDHE-ECDSA-AES256-SHA', 'ECDHE-RSA-AES256-SHA', 'DHE-RSA-AES256-SHA', 'DHE-DSS-AES256-SHA', 'DHE-RSA-CAMELLIA256-SHA', 'DHE-DSS-CAMELLIA256-SHA', 'ECDHE-ECDSA-AES128-SHA', 'ECDHE-RSA-AES128-SHA', 'DHE-RSA-AES128-SHA', 'DHE-DSS-AES128-SHA', 'DHE-RSA-CAMELLIA128-SHA', 'DHE-DSS-CAMELLIA128-SHA', 'AES256-GCM-SHA384', 'AES256-CCM8', 'AES256-CCM', 'ARIA256-GCM-SHA384', 'AES128-GCM-SHA256', 'AES128-CCM8', 'AES128-CCM', 'ARIA128-GCM-SHA256', 'AES256-SHA256', 'CAMELLIA256-SHA256', 'AES128-SHA256', 'CAMELLIA128-SHA256', 'AES256-SHA', 'CAMELLIA256-SHA', 'AES128-SHA', 'CAMELLIA128-SHA']
```

(Tested with OpenSSL 1.1.0c, 1.1.0 / 1.0.2 are similar)

Reply:
D'oh, context matters :-(

Right, I'd be ok limiting this but I don't know if it falls into the purview of this change. It needs to be done upstream first and this particular patch is the ipa-4-6 version of an existing change.